src/XrdVoms/XrdVomsMapfile.cc

```diff
@@ -0,0 +1,417 @@
+        return VOMS_MAP_FAILED;
+    }
+    stream.Attach(cfg_fd);
+    char *var;
+    std::string map_filename;
+    while ((var = stream.GetMyFirstWord())) {
+        if (!strcmp(var, "voms.mapfile")) {
+            auto val = stream.GetWord();
+            if (!val || !val[0]) {
+                if (erp) erp->Emsg("Config", "VOMS mapfile not specified");
+                return VOMS_MAP_FAILED;
+            }
+            map_filename = val;
+        } else if (!strcmp(var, "voms.trace")) {
+            auto val = stream.GetWord();
+            if (!val || !val[0]) {
+                if (erp) erp->Emsg("Config", "VOMS logging level not specified");
+                return VOMS_MAP_FAILED;
+            }
+            if (erp) erp->setMsgMask(0);
+            if (erp) do {
+                if (!strcmp(val, "all")) {erp->setMsgMask(erp->getMsgMask() | LogMask::All);}
+                else if (!strcmp(val, "error")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Error);}
+                else if (!strcmp(val, "warning")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Warning);}
+                else if (!strcmp(val, "info")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Info);}
+                else if (!strcmp(val, "debug")) {erp->setMsgMask(erp->getMsgMask() | LogMask::Debug);}
+                else if (!strcmp(val, "none")) {erp->setMsgMask(0);}
+                else {erp->Emsg("Config", "voms.trace encountered an unknown directive:", val);}
+                val = stream.GetWord();
+            } while (val);
+        }
+    }
+
+    if (!map_filename.empty()) {
+        if (erp) erp->Emsg("Config", "Will initialize VOMS mapfile", map_filename.c_str());
+        mapper.reset(new XrdVomsMapfile(erp, map_filename));
+        if (!mapper->IsValid()) {
+            mapper.reset(nullptr);
+            return VOMS_MAP_FAILED;
+        }
+    }
+
+    return mapper.get();
+}
+
+
+void *
+XrdVomsMapfile::MaintenanceThread(void *myself_raw)
+{
+    auto myself = static_cast<XrdVomsMapfile*>(myself_raw);
+
+   auto now = monotonic_time_s();
+   auto next_update = now + m_update_interval;
+   while (true) {
+       now = monotonic_time_s();
+       auto remaining = next_update - now;
+       auto rval = sleep(remaining);
+       if (rval > 0) {
+           // Woke up early due to a signal; re-run prior logic.
+           continue;
+       }
+       next_update = monotonic_time_s() + m_update_interval;
+       struct stat statbuf;
+       if (-1 == stat(myself->m_mapfile.c_str(), &statbuf)) {
+           myself->m_edest->Emsg("XrdVomsMapfile", errno, "Error checking the mapfile",
+               myself->m_mapfile.c_str());
+           myself->m_mapfile_ctime.tv_sec = 0;
+           myself->m_mapfile_ctime.tv_nsec = 0;
+           myself->m_is_valid = false;
+           continue;
+       }
+       if ((myself->m_mapfile_ctime.tv_sec == statbuf.st_ctim.tv_sec) &&
+           (myself->m_mapfile_ctime.tv_nsec == statbuf.st_ctim.tv_nsec))
```

Comment:
Are you sure you want ctime? I suppose you are banking on the fact that the file will be effectively recreated with say an "mv", right? Can we depend on that or should be ignore changes unless that has actually happened. Either way a comment would be appreciated to that effect.

Reply:
Yup - `ctime` can't be manipulated by userspace tools, making it more reliable for detecting changes rolled out by some external tooling or renames.  Added the comment on this!